ibis/backends/tests/conftest.py

```diff
@@ -0,0 +1,308 @@
+    """Return all backend tests classes defined in thebackends directories."""
+    for backend_dir in Path(__file__).parent.parent.iterdir():
+        backend = backend_dir.name
+        try:
+            conftest = importlib.import_module(
+                f'ibis.backends.{backend}.tests.conftest'
+            )
+        except ImportError:
+            pass
+        else:
+            for obj_name in dir(conftest):
+                obj = getattr(conftest, obj_name)
+                if (
+                    inspect.isclass(obj)
+                    and issubclass(obj, BackendTest)
+                    and obj.name() == backend
+                ):
+                    yield obj
+
+
+ALL_BACKENDS = sorted(
+    backend_test_classes(), key=operator.attrgetter("__name__")
+)
+
+
+def pytest_runtest_call(item):
+    """Dynamically add various custom markers."""
+    nodeid = item.nodeid
+    for marker in list(item.iter_markers(name="only_on_backends")):
+        (backend_types,) = map(tuple, marker.args)
+        backend = item.funcargs["backend"]
+        assert isinstance(
+            backend, BackendTest
+        ), "backend has type {!r}".format(type(backend).__name__)
+        if not isinstance(backend, backend_types):
+            pytest.skip(
+                f"only_on_backends: {backend} is not in {backend_types} "
+                f"{nodeid}"
+            )
+
+    for marker in list(item.iter_markers(name="skip_backends")):
+        (backend_types,) = map(tuple, marker.args)
+        backend = item.funcargs["backend"]
+        assert isinstance(
+            backend, BackendTest
+        ), "backend has type {!r}".format(type(backend).__name__)
+        if isinstance(backend, backend_types):
+            pytest.skip(f"skip_backends: {backend} {nodeid}")
+
+    for marker in list(item.iter_markers(name="skip_missing_feature")):
+        backend = item.funcargs["backend"]
+        (features,) = marker.args
+        missing_features = [
+            feature for feature in features if not getattr(backend, feature)
+        ]
+        if missing_features:
+            pytest.mark.skip(
+                'Backend {} is missing features {} needed to run {}'.format(
+                    type(backend).__name__, ', '.join(missing_features), nodeid
+                )
+            )
+
+    for marker in list(item.iter_markers(name="xfail_backends")):
+        (backend_types,) = map(tuple, marker.args)
+        backend = item.funcargs["backend"]
+        assert isinstance(
+            backend, BackendTest
+        ), "backend has type {!r}".format(type(backend).__name__)
+        item.add_marker(
+            pytest.mark.xfail(
+                condition=isinstance(backend, backend_types),
+                reason='Backend {} does not pass this test'.format(
+                    type(backend).__name__
+                ),
+                **marker.kwargs,
+            )
+        )
+
+    for marker in list(item.iter_markers(name="xpass_backends")):
```

Comment:
we use this?

Reply:
Not really sure. I opened #2570, and will have a look next week. I don't want to remove of change any code in this PR, and just have the code copied as is, so we can track the changes easier later.

Reply:
ok cool